src/ir/child-typer.h

```diff
@@ -0,0 +1,1074 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef wasm_ir_child_typer_h
+#define wasm_ir_child_typer_h
+
+#include "wasm-traversal.h"
+#include "wasm.h"
+
+namespace wasm {
+
+// For each child, call `noteSubtype` with a pointer to the child and the most
+// specific type that the child must have.
+template<typename Subtype> struct ChildTyper : OverriddenVisitor<Subtype> {
+  Module& wasm;
+  Function* func;
+
+  ChildTyper(Module& wasm, Function* func) : wasm(wasm), func(func) {}
+
+  Subtype& self() { return *static_cast<Subtype*>(this); }
+
+  void note(Expression** childp, Type type) {
+    self().noteSubtype(childp, type);
+  }
+
+  void notePointer(Expression** ptrp, Name mem) {
+    note(ptrp, wasm.getMemory(mem)->indexType);
+  }
+
+  void noteAny(Expression** childp) { self().noteAnyType(childp); }
+
+  void noteAnyReference(Expression** childp) {
+    self().noteAnyReferenceType(childp);
+  }
+
+  void noteAnyTuple(Expression** childp, size_t arity) {
+    self().noteAnyTupleType(childp, arity);
+  }
+
+  Type getLabelType(Name label) { return self().getLabelType(label); }
+
+  void visitNop(Nop* curr) {}
+
+  void visitBlock(Block* curr) {
+    size_t n = curr->list.size();
+    for (size_t i = 0; i < n - 1; ++i) {
+      note(&curr->list[i], Type::none);
+    }
+    if (n > 0) {
+      note(&curr->list.back(), curr->type);
+    }
+  }
+
+  void visitIf(If* curr) {
+    note(&curr->condition, Type::i32);
+    note(&curr->ifTrue, curr->type);
+    if (curr->ifFalse) {
+      note(&curr->ifFalse, curr->type);
+    }
+  }
+
+  void visitLoop(Loop* curr) { note(&curr->body, curr->type); }
+
+  void visitBreak(Break* curr, std::optional<Type> labelType = std::nullopt) {
+    if (!labelType) {
+      labelType = getLabelType(curr->name);
```

Comment:
Why not always do this to get the label's type?

Reply:
Sometimes it is more convenient to pass a type explicitly to avoid having to maintain a persistent mapping from labels to types that the callback can use.

For the IRBuilder case, `getLabelType` is not readily implementable, arguably due to a layering issue. IRBuilder uses indices (like in the binary format) rather than names as the source of truth for labels. It can convert label names into label indices, but only for label names that are explicitly provided at the start of a scope, not for label names that are generated on demand because a branch targets an unnamed label. That means that an implementation of this `getLabelType` callback would fail to find the type of labels with generated names.

Fixing this would not be as simple as letting IRBuilder translated generated label names into indices because the parser currently uses IRBuilder's name-to-index translation to check whether a name is valid. If the translation worked for generated names, the parser would start admitting modules that happen to use the generated names without them being bound. This is arguably a layering violation if you assume the parser should be responsible for tracking its own mapping of label names to indices, but on the other hand the entire point of IRBuilder is to track state so the parser doesn't have to.

The current solution is to have IRBuilder always pass an explicit type rather than depending on the `getLabelType` callback. Another solution would be to differentiate generated names from explicit names so that we could have different behavior in the public and private mapping of label names to indices.

Reply:
I see, thanks. Yeah, this seems tricky to fix.

I'd suggest adding a comment summarizing some of what you said here, for the benefit of future readers of the code here (though hopefully we'll find a nice refactoring before long).